### PR TITLE
feat(acvm)!: Support stepwise execution of ACIR

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Janky implementation of stepwise execution.

## Summary\*

This PR makes the assumption that each opcode can be executed on the first attempt (excluding foreign calls). We can then implement stepwise execution.

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: acvm/tests/solver.rs [hunk 1]
```diff
@@ -286,13 +281,12 @@ fn double_inversion_brillig_oracle() {
...
-    assert_eq!(
-        solver_status,
-        PartialWitnessGeneratorStatus::RequiresForeignCall,
+    let solver_status = acvm.solve();
+    assert!(
+        matches!(solver_status, ACVMStatus::RequiresForeignCall(_)),
         "Should require oracle data"
     );
-    assert!(acvm.unresolved_opcodes().is_empty(), "should be fully solved");
+    assert_eq!(acvm.instruction_pointer(), 0, "should be fully solved");
```
Comment (TomAFrench) [inline acvm/tests/solver.rs:289]: assert messages are no longer accurate.

File: acvm/src/pwg/mod.rs [hunk 2]
```diff
@@ -85,66 +89,102 @@ pub enum OpcodeResolutionError {
...
     }
 
+    /// Updates the current status of the VM.
+    /// Returns the given status.
+    fn status(&mut self, status: ACVMStatus) -> ACVMStatus {
+        self.status = status.clone();
+        status
+    }
+
+    /// Sets the status of the VM to `ForeignCallWait`.
```
Comment (TomAFrench) [inline acvm/src/pwg/mod.rs:124]:
```suggestion
    /// Sets the status of the VM to `RequiresForeignCall`.
```

**Comment by kevaundray** [comment]
Is this dependent on #433

**Comment by TomAFrench** [comment]
> Is this dependent on #433

Only to create a new release and have Noir consume it.

File: acir/src/circuit/mod.rs [hunk 3]
```diff
@@ -20,6 +20,8 @@ pub struct Circuit {
     pub current_witness_index: u32,
     pub opcodes: Vec<Opcode>,
 
+    /// The set of private inputs to the circuit.
+    pub private_parameters: BTreeSet<Witness>,
```
Comment (TomAFrench) [inline acir/src/circuit/mod.rs:24]: @kevaundray This will cause the barretenberg codegen to need to be refreshed.

File: acvm/src/compiler/mod.rs [hunk 4]
```diff
@@ -101,26 +110,123 @@ pub fn compile(
...
+                match directive {
+                    Directive::Invert { result, .. } => {
+                        transformer.mark_solvable(*result);
+                    }
+                    Directive::Quotient(quotient_directive) => {
+                        transformer.mark_solvable(quotient_directive.q);
+                        transformer.mark_solvable(quotient_directive.r);
+                    }
+                    Directive::ToLeRadix { b, .. } => {
+                        for w in b {
```
Comment (kevaundray) [inline acvm/src/compiler/mod.rs:172]: Not a fan of one letter variables in most cases; its unavoidable for `b` but can you specify what `w` is?

Comment (kevaundray) [reply acvm/src/compiler/mod.rs:172]: This is done in a lot of places -- we can create an issue to change it everywhere, in directives too, and leave it in this PR

File: acvm/src/compiler/transformers/csat.rs [hunk 5]
```diff
@@ -318,29 +373,37 @@ impl CSatTransformer {
...
+                    intermediate_variables,
+                    intermediate_gate,
+                    num_witness,
+                );
+                self.mark_solvable(inter_var.1);
+                added.push(inter_var);
+            }
+            // The intermediate gate is not full, but the gate still has too many terms
+            if not_full && gate.linear_combinations.len() > self.width {
+                unreachable!("Could not reduce the expression");
```
Comment (kevaundray) [inline acvm/src/compiler/transformers/csat.rs:400]: can you add into the unreachable message, what the expression is?

File: acvm/src/pwg/block.rs [hunk 6]
```diff
@@ -159,7 +120,9 @@ mod tests {
         let mut block_solver = BlockSolver::default();
         block_solver.init(&init, &initial_witness).unwrap();
 
-        block_solver.solve(&mut initial_witness, &trace).unwrap();
+        for op in trace {
+            block_solver.solve_memory_op(&op, &mut initial_witness).unwrap();
```
Comment (kevaundray) [inline acvm/src/pwg/block.rs:124]: This unwrap is more okay since its in a test

File: acvm/src/pwg/mod.rs [hunk 7]
```diff
@@ -154,11 +141,22 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
...
+
+    /// Returns the index of the current opcode to be executed.
+    pub fn instruction_pointer(&self) -> usize {
+        self.instruction_pointer
+    }
+
+    /// Finalize the ACVM execution, returning the resulting [`WitnessMap`].
+    pub fn finalize(self) -> WitnessMap {
+        if self.status != ACVMStatus::Solved {
+            panic!("ACVM is not ready to be finalized");
```
Comment (kevaundray) [inline acvm/src/pwg/mod.rs:157]: Can you add the status into the panic message context?

File: acvm/src/pwg/mod.rs [hunk 8]
```diff
@@ -174,151 +172,122 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
...
+                    self.status(ACVMStatus::Solved)
+                } else {
+                    self.status(ACVMStatus::InProgress)
+                }
+            }
+            Ok(OpcodeResolution::InProgress) => {
+                unreachable!("Opcodes_and_labels should be immediately solvable");
+            }
+            Ok(OpcodeResolution::InProgressBrillig(_)) => {
+                unreachable!("Handled above")
```
Comment (kevaundray) [inline acvm/src/pwg/mod.rs:264]: Can you add more context to this unreachable message? I'm imaginging what happens when I trigger this in Noir for example and theres not much context

File: acvm/src/pwg/mod.rs [hunk 9]
```diff
@@ -174,151 +172,122 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
...
+            Ok(OpcodeResolution::Solved) => {
+                self.instruction_pointer += 1;
+                if self.instruction_pointer == self.opcodes.len() {
+                    self.status(ACVMStatus::Solved)
+                } else {
+                    self.status(ACVMStatus::InProgress)
+                }
+            }
+            Ok(OpcodeResolution::InProgress) => {
+                unreachable!("Opcodes_and_labels should be immediately solvable");
```
Comment (kevaundray) [inline acvm/src/pwg/mod.rs:261]: Can we add more context to this message? Its not immediate clear what "Opcodes_and_labels" means here if looking at it from a dependent package like Noir

**Review by kevaundray — COMMENTED** [review]
Left some comments

**Comment by kevaundray** [comment]
Hmm seems I left comments and github didn't save them 🤔

File: acvm/src/pwg/mod.rs (hunk 9 above)
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:261]: Addressed in https://github.com/noir-lang/acvm/pull/400

File: acvm/src/pwg/mod.rs (hunk 8 above)
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:264]: Addressed in https://github.com/noir-lang/acvm/pull/400

**Review by kevaundray — APPROVED** [review]
Spoke with Tom on slack, we can deal with these nits in a separate PR

File: acvm/src/pwg/mod.rs [hunk 10]
```diff
@@ -174,151 +172,122 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
...
             }
-
-            // Before potentially ending execution, we must save the list of opcodes which remain to be solved.
-            std::mem::swap(&mut self.opcodes_and_labels, &mut unresolved_opcodes);
-
-            // We have oracles that must be externally resolved
-            if self.get_pending_foreign_call().is_some() {
-                return self.status(ACVMStatus::RequiresForeignCall);
+            Opcode::Block(_) | Opcode::ROM(_) | Opcode::RAM(_) => {
+                panic!("Block, ROM and RAM opcodes are not supported by stepwise ACVM")
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:248]: I'm inclined to leave this as is. We should remove these opcodes in the very near future anyway.

File: acvm/src/pwg/mod.rs (hunk 7 above)
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:157]: Addressed in #455

File: acvm/src/pwg/mod.rs [hunk 11]
```diff
@@ -174,151 +172,122 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
...
-            .insert(0, (Opcode::Brillig(resolved_brillig), self.pending_brillig_label_maps[&hash]));
+        let opcode = &mut self.opcodes[self.instruction_pointer];
+        if let Opcode::Brillig(brillig) = opcode {
+            // Overwrite the brillig opcode with a new one with the foreign call response.
+            brillig.foreign_call_results.push(foreign_call_result);
+
+            // Now that the foreign call has been resolved then we can resume execution.
+            self.status(ACVMStatus::InProgress);
+        } else {
+            panic!("Brillig resolution for non brillig opcode");
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:202]: Addressed in https://github.com/noir-lang/acvm/pull/455#discussion_r1274839227

File: acvm/src/pwg/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (TomAFrench) [reply acvm/src/pwg/mod.rs:248]: Addressed in https://github.com/noir-lang/acvm/pull/457

**Pull request merged**